File: android/app/src/main/java/ai/thetahealth/mirobody/data/health/FhirObservation.kt
package ai.thetahealth.mirobody.data.health

import kotlinx.serialization.json.addJsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonArray
import kotlinx.serialization.json.putJsonObject
import java.time.Instant
import java.time.format.DateTimeFormatter

/**
 * Turns a normalized [HealthSample] into a FHIR R4 `Observation` resource, the
 * shape the server's /fhir endpoint validates and stores (src/fhir, whose write
 * model mirrors Android Health Connect). The server assigns the id and meta, so
 * we send only the clinical content.
 */
object FhirObservation {

    /** ISO-8601 UTC instant, e.g. 2026-06-23T14:00:00Z. */
    private fun iso(millis: Long): String =
        DateTimeFormatter.ISO_INSTANT.format(Instant.ofEpochMilli(millis))

    fun from(s: HealthSample) = buildJsonObject {
        put("resourceType", "Observation")
        put("status", "final")
        putJsonArray("category") {
            addJsonObject {
                putJsonArray("coding") {
                    addJsonObject {
                        put("system", "http://terminology.hl7.org/CodeSystem/observation-category")
                        put("code", s.metric.category)
                    }
                }
            }
        }
        putJsonObject("code") {
            putJsonArray("coding") {
                addJsonObject {
                    put("system", "http://loinc.org")
                    put("code", s.metric.loinc)
                    put("display", s.metric.display)
                }
            }
        }
        // Instantaneous samples carry a single effectiveDateTime; interval samples
        // (a day's steps, a sleep session) carry an effectivePeriod.
        if (s.startMillis == s.endMillis) {
            put("effectiveDateTime", iso(s.startMillis))
        } else {
            putJsonObject("effectivePeriod") {
                put("start", iso(s.startMillis))
                put("end", iso(s.endMillis))
            }
        }
        putJsonObject("valueQuantity") {
            put("value", s.value)
            put("unit", s.metric.unit)
            put("system", "http://unitsofmeasure.org")
            put("code", s.metric.ucumUnit)
        }
        // Record the on-device origin (Health Connect / HMS) for provenance.
        putJsonObject("method") {
            putJsonArray("coding") {
                addJsonObject { put("display", s.source) }
            }
        }
    }
}

File: ios/Mirobody/Data/Health/FhirObservation.swift
import Foundation

/// Turns a normalized ``HealthSample`` into a FHIR R4 `Observation` JSON body, the
/// shape the server's /fhir endpoint validates and stores. The server assigns the
/// id and meta, so only the clinical content is sent. Mirrors the Android
/// `FhirObservation`.
enum FhirObservation {

    private static let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    /// Serialize `sample` to a FHIR Observation JSON body.
    static func data(from sample: HealthSample) throws -> Data {
        var observation: [String: Any] = [
            "resourceType": "Observation",
            "status": "final",
            "category": [
                ["coding": [[
                    "system": "http://terminology.hl7.org/CodeSystem/observation-category",
                    "code": sample.metric.category,
                ]]],
            ],
            "code": [
                "coding": [[
                    "system": "http://loinc.org",
                    "code": sample.metric.loinc,
                    "display": sample.metric.display,
                ]],
            ],
            "valueQuantity": [
                "value": sample.value,
                "unit": sample.metric.unit,
                "system": "http://unitsofmeasure.org",
                "code": sample.metric.ucumUnit,
            ],
            "method": [
                "coding": [["display": sample.source]],
            ],
        ]

        // Instantaneous vs interval timing.
        if sample.start == sample.end {
            observation["effectiveDateTime"] = iso.string(from: sample.start)
        } else {
            observation["effectivePeriod"] = [
                "start": iso.string(from: sample.start),
                "end": iso.string(from: sample.end),
            ]
        }

        return try JSONSerialization.data(withJSONObject: observation, options: [])
    }
}
